pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/OffHeapStringStore.java

```diff
@@ -0,0 +1,241 @@
+        endOffset = _byteBuffer.getInt((index - 1) * INT_SIZE);
+      }
+      if ((endOffset - startOffset)/CHAR_SIZE != string.length()) {
+        return false;
+      }
+      for (int i = 0, j = startOffset; i < string.length(); i++, j = j + CHAR_SIZE) {
+        if (string.charAt(i) != _byteBuffer.getChar(j)) {
+          return false;
+        }
+      }
+      return true;
+    }
+
+    private String get(final int index) {
+      int startOffset = _byteBuffer.getInt(index * INT_SIZE);
+      int endOffset = _byteBuffer.capacity();
+      if (index > 0) {
+        endOffset = _byteBuffer.getInt((index - 1) * INT_SIZE);
+      }
+      char[] chars = new char[(endOffset - startOffset)/CHAR_SIZE];
+      for (int i = 0, j = startOffset; i < chars.length; i++, j = j + CHAR_SIZE) {
+        chars[i] = _byteBuffer.getChar(j);
+      }
+      return new String(chars);
+    }
+
+    @Override
+    public void close()
+        throws IOException {
+      _pinotDataBuffer.close();
+    }
+
+    private long getSize() {
+      return _size;
+    }
+
+    private int getStartIndex() {
+      return _startIndex;
+    }
+  }
+
+  private volatile List<Buffer> _buffers = new ArrayList<Buffer>();
+  private int _numElements = 0;
+  private volatile Buffer _currentBuffer;
+
+  public OffHeapStringStore() {
+    expand(START_SIZE);
+  }
+
+  // Expand the buffer size, allocating a min of 32k for strings.
+  private Buffer expand(long size) {
+    Buffer buffer = new Buffer(size, _numElements);
+    List<Buffer> newList = new LinkedList<>();
+    for (Buffer b : _buffers) {
+      newList.add(b);
+    }
+    newList.add(buffer);
+    _buffers = newList;
+    _currentBuffer = buffer;
+    return buffer;
+  }
+
+  private Buffer expand() {
+    Buffer newBuffer = expand(_currentBuffer.getSize() * 2);
+    return newBuffer;
+  }
+
+  // Returns a string, given an index
+  public String get(int index) {
+    List<Buffer> bufList = _buffers;
+    for (int x = bufList.size()-1; x >= 0; x--) {
+      Buffer buffer = bufList.get(x);
+      if (index >= buffer.getStartIndex()) {
+        return buffer.get(index - buffer.getStartIndex());
+      }
+    }
+    // Assumed that we will never ask for an index that does not exist.
+    throw new RuntimeException("dictionary ID '" + index + "' too low");
+  }
+
+  // Adds a string and returns the index. No verification is made as to whether the string already exists or not
+  public int add(String string) {
+    Buffer buffer = _currentBuffer;
+    int index = buffer.add(string);
+    while (index < 0) {
+      buffer = expand();
+      index = buffer.add(string);
+    }
+    _numElements++;
+    return index + buffer.getStartIndex();
+  }
+
+  public boolean equalsStringAt(String string, int index) {
+    List<Buffer> bufList = _buffers;
+    for (int x = bufList.size()-1; x >= 0; x--) {
+      Buffer buffer = bufList.get(x);
+      // Assumed that we will never ask for a
+      if (index >= buffer.getStartIndex()) {
```

Comment:
We assume that this never happens. The general approach has been (in Pinot) to not add defensive programming in lower layers for performance reasons. You can see that pattern in many places in pinot.